### PR TITLE
core: scan a copy-on-write snapshot in GetRegionSizeByRange

### PR DESCRIPTION
**Pull request by vldmit**
### What problem does this PR solve?

Issue Number: ref #9574

**Problem.** While a store is `Preparing`, `checkStores` calls `getThreshold` for it every
`nodeStateCheckJobInterval`. With placement rules that split the keyspace, that reaches
`GetRegionSizeByRange` with a non-empty range, which walks the region tree in
`ScanRegionLimit`-sized chunks and **releases and immediately re-acquires `RegionsInfo.t`
between them**. On a cluster with 1.5M regions one call therefore takes the lock about 1500
times. Go's `sync.RWMutex` is write-preferring, so each of those acquisitions lets at most one
waiting writer through, and region-heartbeat writes are held near that rate for as long as the
stores stay in `Preparing`.

This is the part of #9574 that the two merged changes do not reach. As @lhy1024 summarised on
that issue, #9580 removed the full-range case and #11072 removed the
`Preparing stores x matching rules` amplification, but "each unique range is still scanned in
O(N) time once per round".

**This PR takes a different angle from the two options listed there.** It does not make the
query O(log N), and it does not reduce the CPU spent walking. It removes the *lock coupling*
instead: the walk runs on a snapshot with no lock held at all. That is deliberate, because the
harmful part of the walk is not its CPU cost — during the incident that motivated this work PD
was using 5-10 of 64 cores with no GC pressure — but the ~1500 lock hand-offs it forces on the
heartbeat writers. An O(log N) weighted aggregate is still worth doing and is complementary;
note though that maintaining a range-size aggregate means updating a root-to-leaf sum whenever
`approximate_size` changes, which puts O(log n) work on the region-heartbeat path, whereas this
change puts none there.

**Why not chunk the other scans as well.** #9671 tried that for `ScanRegions` and #9844 reverted
it, to preserve "view consistency of the region between two scans". That objection is right and
it is inherent to chunking: re-entering the tree at the last end key lets the key space change
between chunks. A copy-on-write snapshot has no second scan to be inconsistent with, so it gives
a bounded lock hold *and* a consistent key-range view.

It also fixes a small correctness bug. The chunked loop resumed each chunk at the previous
region's end key, and `scanRange` resolves a key to the region *containing* it, so a merge across
a chunk boundary made the merged region contribute its whole size to a range that had already
been partly counted.

### What is changed and how does it work?

Three steps, one per commit, plus benchmarks and tests.

**1. `regionItem` holds its `*RegionInfo` in an `atomic.Pointer`.**
`setRegionLocked` replaces that pointer while other goroutines may read the item. That is safe
only while every reader holds `RegionsInfo.t`. `atomic.Pointer` is one word, so `regionItem`
stays 8 bytes and there is no per-region memory cost. Dropping the embedded field also removes
the shadowing of `RegionInfo`'s `GetID`, `GetStartKey` and `GetEndKey` by the identically named
methods on `regionItem`.

**2. A fresh `regionItem` on range change.** An item that is in a tree must never change its key
range, because its keys are what orders it in the btree: a reader holding a reference across the
change would see it at one position while it reports another range. `setRegionLocked` now
allocates a new item on the range-changed path and repoints `r.regions`. That path already does
`tree.remove` plus `overlaps`, `Delete` and `ReplaceOrInsert`, so one 8-byte allocation is not
measurable against it. The same-range path is untouched and still replaces the value in place,
which is what a region heartbeat does. The reference count is unchanged.

**3. `rootRangeSnapshot`, and `GetRegionSizeByRange` walks one.**
`pkg/btree` is already copy-on-write and its `Clone` is O(1), but nothing called it. Taking a
snapshot needs the write lock, because `Clone` rewrites the source tree's copy-on-write context,
but that critical section is O(1) regardless of region count.

**Consistency contract, stated plainly.** A snapshot freezes the tree's *shape* — which items
exist, their key ranges, the coverage, the order, the length. It does **not** freeze the values:
the `RegionInfo` behind an item may still be replaced concurrently, though only by one covering
the same key range, so a scan sees exactly one region per key. That is **weaker** than a scan
under one read lock, where every region comes from a single instant. The type documents the
difference, and callers that need agreement between two different regions' values rather than
agreement about which key ranges exist must not use a snapshot. A snapshot also takes no
reference on the regions it holds, since that would make it O(n) and `RegionInfo.ref` carries the
functional "already in the subtree" signal.

**`BatchScanRegions` is deliberately left on the read lock.** I tried converting it, because it
is the function #9844 reverted chunking on, and measured a large regression: it has the opposite
shape, short scans at a high call rate, so an exclusive clone per call costs much more than the
read lock it replaces. Eight-region scans went from 3.1us to 20.7us at GOMAXPROCS=4. Only a
shared snapshot invalidated on structural change would suit that caller. A comment in the code
records the measurement so the next reader does not have to rediscover it.

### Measurements

arm64 (Apple M4 Max), 1M regions, `benchstat`, before/after measured back to back in the same
thermal state. **Not yet measured on amd64.**

The path this PR is about — region-heartbeat writers with a partial-range scanner running
concurrently, `BenchmarkRandomSetRegionWithGetRegionSizeByRangeParallel`:

| GOMAXPROCS | before | after | |
|---|---|---|---|
| 1 | 3.159us ± 96% | 1.758us ± 59% | -44.3% (p=0.026) |
| 4 | 12.597us ± 9% | 0.911us ± 34% | **-92.8%** (p=0.002) |
| 16 | 14.601us ± 9% | 1.085us ± 4% | **-92.6%** (p=0.002) |

Cost added elsewhere:

| | before | after | |
|---|---|---|---|
| same-range write (the heartbeat path) | 385.5ns, 0 B, 0 allocs | 384.1ns, 0 B, 0 allocs | ~ (p=0.755, n=10) |
| same-range write, snapshot outstanding | - | ~365ns, 0 B, 0 allocs | no cost |
| structural write (split/merge) | 2.406us, 40 B, 5 allocs | 2.519us, 48 B, 6 allocs | +4.7% (p=0.000) |
| full-tree scan | 11.35ms | 10.81ms | no regression |
| snapshot creation | - | ~95-107ns, 4 allocs | flat from 100k to 4M regions |

The same-range write path is the one that matters for acceptance, and it is unchanged including
allocations, with and without a snapshot outstanding. It never calls `ReplaceOrInsert` or
`Delete`, so it never reaches btree's copy-on-write node copy.

Copy-on-write is paid only by structural writes, and less than I expected: re-taking a snapshot
every 1000 structural writes costs +36 B/op over not having one, because most writes hit paths
the tree already owns.

An item layout with an immutable `startKey`, which would keep the atomic off the btree
comparison path and make the ordering invariant structural, was prototyped and rejected: it
recovered ~1.7% on the structural path but quadrupled its allocation (40 -> 160 B/op, since every
scratch item grows from 8 to 32 bytes), on top of ~36 MB of steady state at 1.5M regions.

### Check List

Tests

- Unit test

New tests, all passing under `-race`:

- `pkg/btree`: `TestBTreeSizeInfoAfterCloneG` and `TestCloneReadWhileWritingOriginalG`. `Clone`
  was covered by one test that compares `Ascend` order and nothing else, while the test that
  checks the per-node `indices` bookkeeping never clones. Nothing exercised `mutableFor`'s
  copying of `indices`, even though `GetAt` and `GetWithIndex` depend on it, and through them
  `GetCountByRange` and `RandomRegions`. Both pass at degree 2, 32 and 64.
- `pkg/core`: `TestRootRangeSnapshotIsolation`, `TestRootRangeSnapshotEquivalence`,
  `TestGetRegionSizeByRangeConcurrent`, `TestRegionItemRangeImmutability`,
  `TestRootRangeSnapshotAdjacencyUnderMerges`, `TestSetRegionRangeChangedRepointsMap`, and extra
  cases on `TestGetRegionSizeByRange` for boundaries inside a region, a range past the last
  region's start key, and an empty range.

The adjacency test uses merges rather than splits on purpose: a split reaches PD as two separate
updates, the shrunk parent and then the new child, so between them the key space really is
missing a range and a scanner is right to report a hole however it locks. The split half of each
cycle is therefore made atomic with respect to taking a snapshot.

Two of the tests fail without the fresh-item allocation, reporting start keys that go backwards,
which is the signature of a shared item that moved in one tree while a clone still holds it at
its old position.

Benchmark hygiene, in the first commit, so the before/after above compares like with like:

- `BenchmarkRandomSetRegionWithGetRegionSizeByRange` and its `Parallel` variant scanned an
  **empty** key range, which is answered from `regionTree.totalSize` in O(1) and never reaches
  the range scan. Both now scan a partial range.
- Their background reader goroutines never stopped, so they outlived their own benchmark and
  perturbed every later one in the same binary.
- `BenchmarkRandomSetRegion` mutates an already published `*RegionInfo` and re-reports the same
  pointer, so `origin == region` and `updateStat` adds and subtracts the same values. It does not
  model a heartbeat, so `BenchmarkSetRegionSameRange` was added, alternating two immutable
  same-range versions of every region.

Code changes

- No configuration change, no HTTP API change, no persistent data change.

Side effects

- Increased code complexity: a new read-only view type with a documented consistency contract.
- Possible performance regression: +4.7% and +8 B/op on the structural write path
  (split/merge/create/delete). The dominant same-range heartbeat path is unchanged. See
  Measurements.

Related changes

- Need to cherry-pick to the release branch: to be decided. Note #9580 and #11072 are also not
  yet on release-8.5, and this change sits on top of them.

### Release note

```release-note
Fix region heartbeat processing being throttled while stores are in the Preparing state, by
scanning a copy-on-write snapshot of the Region tree instead of repeatedly acquiring the Region
tree lock. This also fixes a Region size that could be over-counted when a Region merge happened
during the scan.
```

## Summary by CodeRabbit

* **Performance**
  * Improved region range queries through consistent read-only snapshots.
  * Enabled more efficient concurrent scans and snapshot-based operations.

* **Reliability**
  * Improved consistency when regions change range, split, merge, or are removed.
  * Strengthened snapshot isolation during concurrent updates.
  * Expanded validation for indexed tree access and copy-on-write behavior.

### CODE REVIEW

File: pkg/core/region_test.go [hunk 1]
```diff
@@ -1085,6 +1375,260 @@ func TestUpdateRegionEquivalence(t *testing.T) {
...
+				size := regions.GetRegionSizeByRange(startKey, endKey)
+				if size < 0 || size > total*2 {
+					t.Errorf("implausible size %d", size)
+					return
+				}
+			}
+		}()
+	}
+
+	// Writers: same-range flow updates, splits, and merges back.
```
Comment (lhy1024) [inline pkg/core/region_test.go:1474]: The writers here do not actually split or merge adjacent Regions. Case 1 only shrinks one Region's end key, and the default case restores it. The assertion also accepts any value up to `total * 2`, so the old chunked implementation could double-count one Region at a merge boundary without failing this test.

`TestRootRangeSnapshotAdjacencyUnderMerges` covers the underlying snapshot range invariant, but it does not exercise the advertised `GetRegionSizeByRange` over-count regression. Could we add a direct regression test that merges adjacent Regions across the former `ScanRegionLimit` boundary, keeps the total size constant, and verifies the returned size? At minimum, this comment should describe the current same-range update and shrink/restore workload.

File: pkg/core/region_tree_test.go [hunk 2]
```diff
@@ -243,10 +244,156 @@ func updateRegions(re *require.Assertions, tree *regionTree, regions []*RegionIn
...
+		id := uint64(i%count + 1)
+		origin := regions.GetRegion(id)
+		re.NotNil(origin)
+		version++
+		var region *RegionInfo
+		if i%2 == 0 {
+			// Same range, new value.
+			region = origin.Clone(SetApproximateSize(int64(i)), SetRegionVersion(version))
+		} else {
+			// Range change: toggle the end key between its original value and a
```
Comment (lhy1024) [inline pkg/core/region_tree_test.go:366]: The range-change branch only changes `EndKey`, while `checkOrdered` verifies only `StartKey` ordering and snapshot length. If a range change reused the existing `regionItem` in place, the snapshot would incorrectly observe the new `EndKey`, but every assertion here could still pass.

`TestRootRangeSnapshotIsolation` covers one deterministic full-range comparison. For this test to serve as the stress guard described above, could it compare complete start/end ranges after each update, or move `StartKey` between deliberately spaced Regions so reusing the old item visibly breaks snapshot ordering?

File: pkg/core/region_tree.go [hunk 3]
```diff
@@ -311,40 +356,120 @@ func (t *regionTree) searchByPrevKeys(prevKeys [][]byte) []*RegionInfo {
...
+// # What a snapshot freezes
+//
+// The tree's shape: the set of items, each item's key range, and therefore the
+// key-space coverage, the iteration order and the length. Regions created, split,
+// merged or removed after the snapshot was taken are invisible to it, and regions
+// removed after it was taken stay visible to it.
+//
+// This is what a chunked scan cannot offer. A scan that releases the lock and
+// re-enters the tree at the last end key can see the key space change underneath
+// it: a split at a chunk boundary yields a region counted twice, and a merge that
```
Comment (lhy1024) [inline pkg/core/region_tree.go:372]: The split and merge outcomes appear to be reversed. If the old chunked scan has counted `[a,b)` and resumes at `b`, merging `[a,b)` and `[b,c)` into `[a,c)` makes `scanRange(b)` return the merged Region, so the already-covered prefix is counted again.

A split does not inherently cause that duplication. Since PD observes the parent shrink and child insertion as separate updates, a scan between them can instead observe a temporary uncovered range. Could we update this description accordingly?
